lib/interfaces/storage.js

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2018 Scott Bender <scott@scottbender.net>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const _ = require('lodash')
+const debug = require('debug')('signalk-server:interfaces:storage')
+const fs = require('fs')
+const path = require('path')
+const jsonpatch = require('json-patch')
+
+const storageUrls = [
+  '/signalk/v1/storage/:appid/:version/*',
+  '/signalk/v1/storage/:appid/:version'
+]
+
```

Comment:
I am a bit uneasy with just relying apps to play nice and not accidentally overwrite each other's data. 

Could we devise a mechanism where (web)apps are injected with their own id and a token that allows them access to their own data? Simple string interpolation for `${APPLICATION_DATA_PATH}` and `${APPLICATION_ACCESS_TOKEN}` when serving index.html?

Or is this too complicated?

Reply:
How would injection work?

We also need to support non webapps.

Reply:
For webapps as outlined above: the app would get the two strings replaced with relevant values in index.html, there you would just grab them like
```
<script>
APPLICATION_DATA_PATH = '${APPLICATION_DATA_PATH}'
</script>

Non-webapps are a different story. The only thing I can think of is that the app provides it's identity in login / access request and we limit access to the application data with the resulting token. But there is nothing keeping one app from masquerading as another, on purpose or accidentally. To prevent accidental mixup we would need to run an app registry.

Maybe not worth the trouble? For webapps the variable init method would allow not hardcoding the path.